gateway/dataimpl/server_v1/xdcrserver.go

```diff
@@ -0,0 +1,849 @@
+	// the document will be overwritten and include the original document content.
+
+	// keep the source
+	return &internal_xdcr_v1.CheckDocumentResponse{}, nil
+}
+
+func (s *XdcrServer) PushDocument(
+	ctx context.Context,
+	in *internal_xdcr_v1.PushDocumentRequest,
+) (*internal_xdcr_v1.PushDocumentResponse, error) {
+	bucketAgent, oboUser, errSt := s.authHandler.GetMemdOboAgent(ctx, in.BucketName)
+	if errSt != nil {
+		return nil, errSt.Err()
+	}
+
+	errSt = s.checkKey(in.Key)
+	if errSt != nil {
+		return nil, errSt.Err()
+	}
+
+	errSt = s.checkCAS(&in.StoreCas)
+	if errSt != nil {
+		return nil, errSt.Err()
+	}
+
+	var docDatatype memdx.DatatypeFlag
+	var docContent []byte
+	if !in.IsDeleted {
+		if in.ContentType == internal_xdcr_v1.ContentType_CONTENT_TYPE_JSON {
+			docDatatype |= memdx.DatatypeFlagJSON
+		}
+
+		switch content := in.Content.(type) {
+		case *internal_xdcr_v1.PushDocumentRequest_ContentUncompressed:
+			docContent = content.ContentUncompressed
+		case *internal_xdcr_v1.PushDocumentRequest_ContentCompressed:
+			docContent = content.ContentCompressed
+			docDatatype |= memdx.DatatypeFlagCompressed
+		default:
+			return nil, status.New(codes.InvalidArgument, "CompressedContent or UncompressedContent must be specified.").Err()
+		}
+
+		var xattrBlob []byte
+		if in.Xattrs != nil {
+			for key, value := range in.Xattrs {
+				xattrBlob = memdx.AppendXattrBlobEntry(xattrBlob, key, string(value))
+			}
+		}
+		if len(xattrBlob) > 0 {
+			// if we need to add xattrs, we need to uncompress the document content first
+			uncompressedContent, errSt := CompressHandler{}.UncompressContent(docContent, docDatatype)
+			if errSt != nil {
+				return nil, errSt.Err()
+			}
+
+			// join the xattr blob and the document content
+			joinedContent := memdx.JoinXattrBlob(xattrBlob, uncompressedContent)
+
+			// now update the document content, and mark it as include xattrs
+			docContent = joinedContent
+			docDatatype &^= memdx.DatatypeFlagCompressed
+			docDatatype |= memdx.DatatypeFlagXattrs
+		}
+	} else {
+		if in.Content != nil {
+			return nil, status.New(codes.InvalidArgument, "Content cannot be specified when IsDeleted is true.").Err()
+		}
+		if in.Xattrs != nil {
+			return nil, status.New(codes.InvalidArgument, "Xattrs cannot be specified when IsDeleted is true.").Err()
+		}
+	}
+
+	crMode, err := bucketAgent.GetConflictResolutionMode(ctx)
+	if err != nil {
+		return nil, s.errorHandler.NewGenericStatus(err).Err()
+	}
+
+	if in.IsDeleted {
+		var checkCas uint64
+		if in.CheckCas != nil {
+			if *in.CheckCas == 0 {
+				return nil, s.errorHandler.NewZeroCasStatus().Err()
+			}
+			checkCas = *in.CheckCas
+		} else {
+			checkCas = 0
+		}
+
+		var opts gocbcorex.DeleteWithMetaOptions
+		opts.OnBehalfOf = oboUser
+		opts.ScopeName = in.ScopeName
+		opts.CollectionName = in.CollectionName
+		opts.Key = []byte(in.Key)
+		opts.RevNo = in.Revno
+		opts.CheckCas = checkCas
+		opts.StoreCas = in.StoreCas
+
+		if checkCas != 0 {
```

Comment:
nit: I'd find this more readable if we had a variable: `skipConflictRsolution := checkCas != 0`, then we could use it in the error handling below also.